ShacShifter/ShapeParser.py

```diff
@@ -35,40 +36,21 @@ def parseShape(self, inputFilePath):
         return self.nodeShapes
 
     def getNodeShapeUris(self):
-        """Get URIs of all Node shapes.
+        """Get URIs of all Root Node shapes.
 
         returns: list of Node Shape URIs
         """
         nodeShapeUris = set()
 
-        for stmt in self.g.subjects(rdflib.RDF.type, self.sh.NodeShape):
-            if stmt not in nodeShapeUris:
-                nodeShapeUris.add(stmt)
+        qres = g.query("""
+            SELECT DISTINCT ?root
+            WHERE {
+                ?root ?s ?o .
+                FILTER NOT EXISTS {?a ?b ?root .}
+            }""")
 
-        for stmt in self.g.subjects(self.sh.property, None):
-            if stmt not in nodeShapeUris:
-                nodeShapeUris.add(stmt)
-
-        for stmt in self.g.subjects(self.sh.targetClass, None):
-            if stmt not in nodeShapeUris:
-                nodeShapeUris.add(stmt)
-
-        for stmt in self.g.subjects(self.sh.targetNode, None):
-            if stmt not in nodeShapeUris:
-                nodeShapeUris.add(stmt)
-
-        for stmt in self.g.subjects(self.sh.targetObjectsOf, None):
-            if stmt not in nodeShapeUris:
-                nodeShapeUris.add(stmt)
-
-        for stmt in self.g.subjects(self.sh.targetSubjectsOf, None):
-            if stmt not in nodeShapeUris:
-                nodeShapeUris.append(stmt)
-
-        # actually not exactly a nodeshape
-        # for stmt in self.g.subjects(rdflib.RDF.type, self.sh.PropertyGroup):
-        #     if not(stmt in nodeShapeUris):
-        #         nodeShapeUris.append(stmt)
+        for row in qres:
+            nodeShapeUris.add(row.root)
 
         return set(list(nodeShapeUris))
```

Comment:
nodeShapeUris is already a list and should not contain duplicates
maybe find a better variable name.